packages/shortest/src/browser/core/snapshot-browser-tool.ts

```diff
@@ -0,0 +1,55 @@
+import { writeFileSync, appendFileSync, mkdirSync, existsSync } from 'fs';
+import { join } from 'path';
+import { BrowserTool } from './browser-tool';
+import { ActionInput, ToolResult } from '../../types/browser';
+import { TestContext } from '../../types/test';
+
+export class SnapshotBrowserTool extends BrowserTool {
+  private snapshotDir: string;
+  private snapshotFile?: string;
+  private shouldRecord: boolean = false;
+
+  constructor(...args: ConstructorParameters<typeof BrowserTool>) {
+    super(...args);
+    
+    // Create snapshots directory
+    this.snapshotDir = join(process.cwd(), '.shortest', 'snapshots');
+    mkdirSync(this.snapshotDir, { recursive: true });
+
+    // Initialize snapshot file when test context is available
+    if (args[2].testContext?.currentTest?.name) {
+      this.initializeSnapshotFile(args[2].testContext.currentTest.name);
+    }
+  }
+
+  private initializeSnapshotFile(testName: string): void {
+    this.snapshotFile = join(this.snapshotDir, `${testName}.test.snapshot.jsonl`);
+    // Only set up recording if snapshot doesn't exist
+    this.shouldRecord = !existsSync(this.snapshotFile);
+    
+    if (this.shouldRecord) {
+      // Create/clear the file only if we're going to record
+      writeFileSync(this.snapshotFile!, '');
+    }
+  }
+
+  async execute(input: ActionInput): Promise<ToolResult> {
```

Comment:
Seems like this part is executing the test steps twice. Once when AI performs it and once as a replay. Perhaps this should be avoided. Maybe the first time we run `pnpm shortest` we should let AI execute its actions and then run the replay in the consecutive time. Would be cool to have a smart mechanism to delete the snapshots once they become obsolete (No rush for now I will add this as a new issue later - unless you want to tackle this in the same PR).

Reply:
I'll try to make something usable inside a flag to keep moving forward.